Hardware/Hardware/CardReader/CardReaderServiceError.swift

```diff
@@ -1,20 +1,152 @@
 /// Models errors thrown by the CardReaderService.
-/// This is doing the bare minimum for now.
-/// Proper error handling is coming in
-/// https://github.com/woocommerce/woocommerce-ios/issues/3734
+/// It identifies the interaction with the card reader
+/// where the error was thrown.
+/// It is important to mark the specific interaction
+/// because some operations (like, for example, procesing a payment)
+/// require of three interactions with the SDK in sequence, and any error
+/// in any of the steps in that sequence would make the whole operation fail
 public enum CardReaderServiceError: Error {
     /// Error thrown during reader discovery
-    case discovery
+    case discovery(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while connecting to a reader
-    case connection
+    case connection(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while creating a payment intent
-    case intentCreation
+    case intentCreation(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while collecting payment methods
-    case paymentMethod
+    case paymentMethod(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while capturing a payment
-    case capturePayment
+    case capturePayment(underlyingError: UnderlyingError = .internalServiceError)
+}
+
+
+/// Underlying error. Models the specific error that made a given
+/// interaction with the SDK fail.
+public enum UnderlyingError: Error {
```

Comment:
I think we should consider separate enums for each "underlying error" for each `CardReaderServiceError` case.  If I am writing discovery and connection logic, I shouldn't need to be aware of the underlying error cases for payment method creation.

Reply:
Consider the case where we are chaining several operations in sequence, for example when processing a payment, which requires to create an intent, collect a payment method and then process the intent.

If we have a single error with an underlying error, we can chain those operations like this (as we are doing in the CardPresentPaymentStore at the moment):

```
        cardReaderService.createPaymentIntent(parameters)
            .flatMap {
                self.cardReaderService.collectPaymentMethod()
            }.flatMap {
                self.cardReaderService.processPayment()
            }.sink { error in
            switch error {
            case .failure(let error):
                onCompletion(.failure(error))
            case .finished:
                onCompletion(.success(()))
            }
        } receiveValue: { intent in
            print("==== Yosemite log for testing. Payment intent after processing payment")
            print(intent)
            print("//// payment intent processed ")
            // TODO. Initiate final step. Update order. Submit intent id to backend.
            // Deferred to https://github.com/woocommerce/woocommerce-ios/issues/3825
            onCompletion(.success(()))
        }.store(in: &cancellables)
```

And then let any layer on top of that decide how to react to that error, because it can be aware of the operation where it failed, and the reason why it failed.

Reply:
> e.g. error will equal CardReaderServiceError.intentCreation(.whateverfailedinthesdk)

I think the downside is if I want to do something with .whateverfailedinthesdk I'll need to cherry pick cases for ifs, right? Or, god forbid, a full switch. Not having all the errors in one enum reduces the number of enums the UI writer needs to consider writing ifs for (or handling in a comprehensive switch)

I'd like to avoid ending up with the user seeing a default case "An unexpected error occurred"-"OK" and then we get HE's hit with that and we don't have a clear idea what they ran in to

Reply:
>  I'd like to avoid ending up with the user seeing a default case "An unexpected error occurred"-"OK" and then we get HE's hit with that and we don't have a clear idea what they ran in to

Absolutely! There is no reason to do that. Still, the way to avoid that is by giving the UI information as granular as possible, right? 😄

Reply:
Yes, and making it as easy as possible to get that granular error information (and any potential corrective action) in front of the users. Looking forward to seeing how this plays out as we continue wiring up the UI.

Thanks!